tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc

```diff
@@ -1351,6 +1351,27 @@ static LogicalResult Verify(BroadcastOp op) {
   return success();
 }
 
+OpFoldResult BroadcastOp::fold(ArrayRef<Attribute> attrs) {
+  auto type = getType().cast<RankedTensorType>();  
+  if (type == getOperand().getType()) {
+    auto sizesSize = broadcast_sizes().getType().getNumElements();
+    if (sizesSize != 0) {
+      return {};
+    }
+    return getOperand();
+  }
+
+  // Constant fold when an operand is a splat tensor attribute.
+  if (!attrs[0] || !type.hasStaticShape()) return {};
+  auto splatOperandAttr = attrs[0].dyn_cast<SplatElementsAttr>();
+  if (!splatOperandAttr) return {};
+  // MLIR core bug (https://bugs.llvm.org/show_bug.cgi?id=46588): dense element
+  // attribute iterator not implemented for complex element types.
```

Comment:
This is fixed already, can you try enabling this and adding a test for it?

Reply:
There is an assertion failed when adding complex type test: 
`mlir/lib/IR/BuiltinAttributes.cpp:688: static mlir::DenseElementsAttr mlir::DenseElementsAttr::get(mlir::ShapedType, ArrayRef<mlir::Attribute>): Assertion `attr.isa<StringAttr>() && "expected string value for non integer/index/float element"' failed.`

Reply:
So I think this should be fixed after fixing `DenseElementsAttr` in MLIR

Reply:
This is just not the right API to use to create a Dense attribute with Complex.
Instead there are multiple methods for it:

```
   // works with T = std::complex<U>
  static DenseElementsAttr get(const ShapedType &type, ArrayRef<T> values) {

  /// Constructs a dense float elements attribute from an array of APFloat
  /// values. Each APFloat value is expected to have the same bitwidth as the
  /// element type of 'type'. 'type' must be a vector or tensor with static
  /// shape.
  static DenseElementsAttr get(ShapedType type, ArrayRef<APFloat> values);

  /// Constructs a dense complex elements attribute from an array of APFloat
  /// values. Each APFloat value is expected to have the same bitwidth as the
  /// element type of 'type'. 'type' must be a vector or tensor with static
  /// shape.
  static DenseElementsAttr get(ShapedType type,
                               ArrayRef<std::complex<APFloat>> values);
```

Reply:
Done, thanks! I also fix the same problem in `BroadcastInDimOp`.